File: seeds/research.py
"""Seed: research — multi-angle gather, adversarial per-claim fact-check, synthesize.

args: a question string, or {"question": "..."}.
"""

META = {
    "name": "research",
    "description": "Answer a research question with sourced, fact-checked claims.",
    "phases": ["Gather", "Verify", "Synthesize"],
}

CLAIM_SCHEMA = {
    "type": "object", "additionalProperties": False,
    "properties": {"claims": {"type": "array", "items": {
        "type": "object", "additionalProperties": False,
        "properties": {"claim": {"type": "string"}, "support": {"type": "string"}},
        "required": ["claim", "support"]}}},
    "required": ["claims"],
}
VERDICT_SCHEMA = {
    "type": "object", "additionalProperties": False,
    "properties": {"supported": {"type": "boolean"}, "why": {"type": "string"}},
    "required": ["supported", "why"],
}


async def run(agent, parallel, pipeline, log, phase, args):
    question = args if isinstance(args, str) else (args or {}).get("question", "")

    phase("Gather")
    angles = ["core definitions & established facts",
              "recent developments & data",
              "counter-evidence, caveats & failure modes"]
    gathered = await parallel([
        (lambda a=a: agent(
            f"Research this angle — '{a}' — for the question:\n{question}\n\n"
            "Return 3-6 concrete factual claims, each with its supporting evidence.",
            label=f"gather:{a[:18]}", phase="Gather", model="sonnet", schema=CLAIM_SCHEMA))
        for a in angles])
    claims = [c for g in gathered if g for c in g.get("claims", [])]
    log(f"gathered {len(claims)} claims")

    phase("Verify")
    async def verify(c):
        v = await agent(
            "Adversarially fact-check this claim. Try to REFUTE it; set supported=false "
            "if the evidence is weak, dated, or over-generalized.\n"
            f"CLAIM: {c['claim']}\nSTATED SUPPORT: {c['support']}",
            label="verify", phase="Verify", model="opus", schema=VERDICT_SCHEMA)
        return {**c, "verdict": v}
    checked = await parallel([(lambda c=c: verify(c)) for c in claims])
    kept = [c for c in checked if c and c.get("verdict") and c["verdict"]["supported"]]
    log(f"{len(kept)}/{len(claims)} claims survived fact-check")

    phase("Synthesize")
    answer = await agent(
        f"Write a concise, well-structured answer to:\n{question}\n\n"
        "Use ONLY these verified claims; do not add unsupported statements:\n"
        + "\n".join(f"- {c['claim']}" for c in kept),
        label="synthesize", phase="Synthesize", model="opus")

    return {"question": question, "answer": answer,
            "claims_verified": len(kept), "claims_total": len(claims)}

File: claude-code/seeds/research.js
// Seed: research — multi-angle gather, adversarial per-claim fact-check, synthesize.
// Claude Code Workflow edition. args: a question string, or { question }.
export const meta = {
  name: 'research',
  description: 'Answer a research question with fact-checked claims: gather from multiple angles, adversarially verify each claim, then synthesize.',
  phases: [{ title: 'Gather' }, { title: 'Verify' }, { title: 'Synthesize' }],
}

const AGENT_TIMEOUT_MS = 240000  // guard: a hung agent resolves null, never stalls parallel()
const aw = (p, o) => Promise.race([
  agent(p, o),
  new Promise((r) => setTimeout(() => { log(`timeout: ${(o && o.label) || 'agent'}`); r(null) }, AGENT_TIMEOUT_MS)),
])

const question = (typeof args === 'string') ? args : (args && args.question) || ''

const CLAIM_SCHEMA = {
  type: 'object', additionalProperties: false, required: ['claims'],
  properties: { claims: { type: 'array', items: {
    type: 'object', additionalProperties: false, required: ['claim', 'support'],
    properties: { claim: { type: 'string' }, support: { type: 'string' } } } } },
}
const VERDICT_SCHEMA = {
  type: 'object', additionalProperties: false, required: ['supported', 'why'],
  properties: { supported: { type: 'boolean' }, why: { type: 'string' } },
}

phase('Gather')
const angles = ['core definitions & established facts', 'recent developments & data',
                'counter-evidence, caveats & failure modes']
const gathered = await parallel(angles.map((a) => () =>
  aw(`Research this angle — '${a}' — for the question:\n${question}\n\n` +
        'Return 3-6 concrete factual claims, each with its supporting evidence.',
        { label: `gather:${a.slice(0, 18)}`, phase: 'Gather', model: 'sonnet', schema: CLAIM_SCHEMA })))
const claims = gathered.filter(Boolean).flatMap((g) => g.claims || [])
log(`gathered ${claims.length} claims`)

phase('Verify')
const checked = await parallel(claims.map((c) => () =>
  aw('Adversarially fact-check this claim. Try to REFUTE it; set supported=false if the ' +
        `evidence is weak, dated, or over-generalized.\nCLAIM: ${c.claim}\nSTATED SUPPORT: ${c.support}`,
        { label: 'verify', phase: 'Verify', model: 'opus', schema: VERDICT_SCHEMA })
    .then((v) => ({ ...c, verdict: v }))))
const kept = checked.filter((c) => c && c.verdict && c.verdict.supported)
log(`${kept.length}/${claims.length} claims survived fact-check`)

phase('Synthesize')
const answer = await aw(
  `Write a concise, well-structured answer to:\n${question}\n\n` +
  'Use ONLY these verified claims; do not add unsupported statements:\n' +
  kept.map((c) => `- ${c.claim}`).join('\n'),
  { label: 'synthesize', phase: 'Synthesize', model: 'opus' })

return { question, answer, claims_verified: kept.length, claims_total: claims.length }
